src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.ClassifiedLineCache.cs

```diff
@@ -0,0 +1,175 @@
+        /// </summary>
+        private readonly Dictionary<Span, LinkedListNode<SpanAndClassifiedSpans>> _spanToLruNode = [];
+
+        private void ClearIfDifferentSnapshot(
+            DocumentId documentId,
+            ParseOptions? parseOptions,
+            ITextSnapshot snapshot)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            if (_documentId != documentId ||
+                _parseOptions != parseOptions ||
+                _snapshot != snapshot)
+            {
+                _lruList.Clear();
+                _spanToLruNode.Clear();
+
+                _documentId = documentId;
+                _parseOptions = parseOptions;
+                _snapshot = snapshot;
+            }
+        }
+
+        public bool TryUseCache(
+            DocumentId documentId,
+            ParseOptions? parseOptions,
+            SnapshotSpan snapshotSpan,
+            SegmentedList<ClassifiedSpan> classifications)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            ClearIfDifferentSnapshot(documentId, parseOptions, snapshotSpan.Snapshot);
+
+            if (!_spanToLruNode.TryGetValue(snapshotSpan.Span, out var node))
+                return false;
+
+            // Move this node to the front of the LRU list.
+            _lruList.Remove(node);
+            _lruList.AddLast(node);
+
+            // AddRange is optimized to take a SegmentedList and copy directly from it into the result list.
+            classifications.AddRange(node.Value.ClassifiedSpans);
+            return true;
+        }
+
+        public void Update(
+            SnapshotSpan snapshotSpan, SegmentedList<ClassifiedSpan> newClassifications)
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            if (newClassifications.Count > MaxClassificationsCount)
+                return;
+
+            var span = snapshotSpan.Span;
+
+            if (_spanToLruNode.ContainsKey(span))
+            {
+                Debug.Fail("This should not be possible.  Caller would have seen this item when calling TryUseCache");
+                return;
+            }
+
+            var node = GetLruNode(span, newClassifications);
+            Contract.ThrowIfTrue(node.Value.Span != span);
+
+            _lruList.AddLast(node);
+            _spanToLruNode.Add(span, node);
+
+            Contract.ThrowIfTrue(_lruList.Count > CacheSize);
+            Contract.ThrowIfTrue(_lruList.Count != _spanToLruNode.Count);
+        }
+
+        private LinkedListNode<SpanAndClassifiedSpans> GetLruNode(Span span, SegmentedList<ClassifiedSpan> newClassifications)
+        {
+            if (_lruList.Count < CacheSize)
+            {
+                // We're not at capacity.  Create a new node to go at the end of the LRU list. Note: The SegmentedList
+                // constructor fast paths the case where we pass in another SegmentedList.
+                return new(new(span, new SegmentedList<ClassifiedSpan>(newClassifications)));
```

Comment:
>new(new(span, new [](http://example.com/codeflow?start=23&length=17)

Just so I know where your line is, this would be too much new?

new(new(span, new(newClassifications)));

Reply:
how about `new(new(s, new(nc)));`?